### PR TITLE
Calculator: Improve factorial/modulo support. IAFD Fixes

### PR DESCRIPTION
**Pull request by pjhampton**
## Description of new Instant Answer, or changes

- Adds factorial support.
- Improves modulo support
- Reduces overtriggering (IAFD from pixel)

## Related Issues and Discussions

Fixes: #1917 
Fixes: #2778

## People to notify

@moollaza 

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/calculator

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Goodie/Calculator.pm [hunk 1]
```diff
@@ -151,19 +158,19 @@ handle query => sub {
 
     # We need to rewrite the functions for front-end consumption
     $query = rewriteFunctions($query);
-
     # throw out obvious non-calculations immediately
     return if $query =~ qr/(\$(.+)?(?=£|€))|(£(.+)?(?=\$|€))|(€(.+)?(?=\$|£))/; # only let one currency type through
     return if $req->query_lc =~ /^0x/i; # hex maybe?
+    return if $query =~ /\d+(?:X|x)\.?$/; # websites such as 1337X
```
Comment (pjhampton) [inline lib/DDG/Goodie/Calculator.pm]: These are the most common overtrigger...

File: lib/DDG/Goodie/Calculator.pm [hunk 2]
```diff
@@ -151,19 +158,19 @@ handle query => sub {
...
+    return if $query =~ /^(?:\.\w+)/; # Probably looking at file extensions
```
Comment (pjhampton) [inline lib/DDG/Goodie/Calculator.pm]: Surprisingly this was overtriggering on a lot of file extensions

File: lib/DDG/Goodie/Calculator.pm [hunk 3]
```diff
@@ -151,19 +158,19 @@ handle query => sub {
...
     return if $query =~ $network;    # Probably want to talk about addresses, not calculations.
     return if $query =~ m/^(\+?\d{1,2}(\s|-)?|\(\d{2})?\(?\d{3,4}\)?(\s|-)?\d{3}(\s|-)?\d{3,4}(\s?x\d+)?$/; # Probably are searching for a phone number, not making a calculation
     return if $query =~ m/(\d+)\s+(\d+)/; # if spaces between numbers then bail
+    return if $query =~ m/^\)|\($/; # shouldn't open with a closing brace or finish with an opening brace
 
     # some shallow preprocessing of the query
-    $query =~ s/(\d+)\s+%\s?(\d+)/$1mod$2/;
```
Comment (pjhampton) [inline lib/DDG/Goodie/Calculator.pm]: I changed my mind on this direction

File: t/Calculator.t [hunk 4]
```diff
@@ -87,9 +87,6 @@ ddg_goodie_test(
     'tanh(1)' => build_test(
         'tanh(1)'
     ),
-    'cotan(1)' => build_test(
-        'cotan(1)'
```
Comment (pjhampton) [inline t/Calculator.t]: we don't support cotan in the front end

File: t/Calculator.t [hunk 5]
```diff
@@ -372,6 +369,33 @@ ddg_goodie_test(
...
+        '8 * 77! + 32'
+    ),
+    '8 * fact77 + 32' => build_test(
+        '8 * 77! + 32'
+    ),
+    'fact(20)' => build_test(
+        '20!'
+    ),
+    'fact 20' => build_test(
+        '20!'
```
Comment (pjhampton) [inline t/Calculator.t]: Will support the bang syntax for the front-end as that is what math.js uses.

**Comment by moollaza** [comment]
@pjhampton Thanks! This LGTM :+1:

**Pull request merged**